shells/android/java/arcs/storage/ServiceStore.kt

```diff
@@ -0,0 +1,175 @@
+    private val coroutineContext: CoroutineContext = Dispatchers.IO,
+    private val connectionFactory: ConnectionFactory? = null
+) {
+    suspend operator fun invoke(
+        options: StoreOptions<Data, Op, ConsumerData>
+    ): ServiceStore<Data, Op, ConsumerData> {
+        val storeContext = coroutineContext + CoroutineName("ServiceStore(${options.storageKey})")
+        return ServiceStore(
+            options = options,
+            crdtType = crdtType,
+            lifecycle = lifecycle,
+            connectionFactory = connectionFactory
+                ?: DefaultConnectionFactory(context, coroutineContext = storeContext),
+            coroutineContext = storeContext
+        ).initialize()
+    }
+}
+
+/** Implementation of [ActiveStore] which pipes [ProxyMessage]s to and from the [StorageService]. */
+@UseExperimental(FlowPreview::class)
+@ExperimentalCoroutinesApi
+class ServiceStore<Data : CrdtData, Op : CrdtOperation, ConsumerData>(
+    private val options: StoreOptions<Data, Op, ConsumerData>,
+    private val crdtType: ParcelableCrdtType,
+    lifecycle: Lifecycle,
+    private val connectionFactory: ConnectionFactory,
+    private val coroutineContext: CoroutineContext
+) : ActiveStore<Data, Op, ConsumerData>(options), LifecycleObserver {
+    private val scope = CoroutineScope(coroutineContext)
+    private var storageService: IStorageService? = null
+    private var serviceConnection: StorageServiceConnection? = null
+    private val proxyCallbacks = ProxyCallbackManager<Data, Op, ConsumerData>()
+    private var serviceCallbackToken: Int = -1
+    private val sendQueue = SendQueue()
+
+    init {
+        lifecycle.addObserver(this)
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override suspend fun getLocalData(): Data {
+        val service = checkNotNull(storageService)
+        val channel = ParcelableProxyMessageChannel(coroutineContext)
+        service.getLocalData(channel)
+        val flow = channel.asFlow()
+        val modelUpdate =
+            flow.onEach { it.result.complete(true) }
+                .mapNotNull {
+                    it.message.actual as? ProxyMessage.ModelUpdate<Data, Op, ConsumerData>
+                }
+                .first()
+        channel.cancel()
+        return modelUpdate.model
+    }
+
+    override fun on(callback: ProxyCallback<Data, Op, ConsumerData>): Int =
+        proxyCallbacks.register(callback)
+
+    override fun off(callbackToken: Int) = proxyCallbacks.unregister(callbackToken)
+
+    override suspend fun onProxyMessage(message: ProxyMessage<Data, Op, ConsumerData>): Boolean {
+        val service = checkNotNull(storageService)
+        val result = DeferredResult(coroutineContext)
+        sendQueue.enqueue {
+            service.sendProxyMessage(message.toParcelable(crdtType), result)
+        }
+        return result.await()
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
+    suspend fun initialize() = apply {
```

Comment:
what's the point of `apply` here?

Reply:
apply returns a reference to the receiver, so setting the function equal to the result of apply makes the function return the ServiceStore itself, which allows chaining.  (Like builder methods typically return the builder)

Reply:
So this is equivalent to:

```
suspend fun initialize() {
  // blah
  return this
}
```

??

Reply:
Yep!